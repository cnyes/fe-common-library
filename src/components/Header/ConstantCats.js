/* eslint-disable import/prefer-default-export */
// prod: http://news.cnyes.cool/api/v3/categories
import CategoriesRawProd from './raw/CategoriesProd.raw.json'; // eslint-disable-line import/no-named-as-default
import CategoriesRawBeta from './raw/CategoriesBeta.raw.json'; // eslint-disable-line import/no-named-as-default

const CategoriesRaw = process.env.NODE_ENV === 'production' ? CategoriesRawProd : CategoriesRawBeta;

const BASE_CATEGORIES = [
  {
    slug: 'all',
    categoryId: 'all',
    name: '總覽',
    parentId: 0,
    parentSlug: '',
    api: '/api/v2/news',
    subs: null,
  },
  {
    slug: 'headline',
    categoryId: 'headline',
    name: '即時頭條',
    parentId: 0,
    parentSlug: '',
    api: '/api/v3/news/category/headline',
    subs: null,
  },
  {
    slug: 'news24h',
    categoryId: 'news24h',
    name: '24HR',
    parentId: 0,
    parentSlug: '',
    api: '/api/v1/news/24h',
    subs: null,
  },
  {
    slug: 'popular',
    categoryId: 'popular',
    name: '人氣',
    parentId: 0,
    parentSlug: '',
    api: '/api/v1/news/popular',
    subs: null,
  },
];
const Categories = [...BASE_CATEGORIES, ...CategoriesRaw.items];

export const CategoryMappingWithSubs = Categories.reduce((pValue, cValue, cIndex, cats) => {
  const cat = cats[cIndex];
  let subs = null;
  let parentSlug = '';

  // build subs
  if (cat.parentId === 0) {
    subs = Categories.reduce((previousValue, currentValue, currentIndex, array) => {
      const subCat = array[currentIndex];

      if (subCat.parentId === cat.categoryId) {
        previousValue.push(subCat.slug);
      }

      return previousValue;
    }, []);

    subs = subs.length > 0 ? subs : null;
  } else {
    // find parent's slug
    const parentCat = Categories.filter(_cat => {
      return _cat.categoryId === cat.parentId;
    });

    if (parentCat && parentCat.length > 0) {
      parentSlug = parentCat[0].slug;
    }
  }

  const result = {
    ...cat,
    parentSlug,
    subs,
  };

  pValue[cat.slug] = result; // eslint-disable-line no-param-reassign

  return pValue;
}, {});
