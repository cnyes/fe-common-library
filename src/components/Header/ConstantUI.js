/* eslint-disable import/prefer-default-export */
export const FIXED_HEADER_NONE = 'FIXED_HEADER_NONE';
export const FIXED_HEADER_FULL = 'FIXED_HEADER_FULL';
export const FIXED_HEADER_SEARCH = 'FIXED_HEADER_SEARCH';
export const FIXED_HEADER_SUB = 'FIXED_HEADER_SUB';
