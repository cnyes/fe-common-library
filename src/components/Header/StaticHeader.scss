@import '../../styles/base';

#cnyes-header-wrapper { // sass-lint:disable-line no-ids
  /* NOTE: reset for export to common component start */
  font-smooth: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font: 400 17px/23px $font-family;
  text-align: left;

  * {
    box-sizing: border-box;
  }

  a,
  a:hover {
    text-decoration: none;
  }
  /* NOTE: reset for export to common component end */
}
