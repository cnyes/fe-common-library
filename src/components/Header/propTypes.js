/* eslint-disable import/prefer-default-export */
import { PropTypes } from 'react';

export const requestType = PropTypes.func;
