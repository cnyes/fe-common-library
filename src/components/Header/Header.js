/* eslint import/no-extraneous-dependencies: ["error", {"peerDependencies": true}] */
import React, { PureComponent, PropTypes } from 'react';
import { Link } from 'react-router';
import { locationShape } from 'react-router/lib/PropTypes';
import raf from 'raf';
import classNames from 'classnames';
import { requestType, navsType, catNavsType } from './propTypes';
import SubMenu from './SubMenu';
import SubNavItem from './SubNavItem';
import styles from './Header.scss';
// import navs from './config/navs';
import { CategoryMappingWithSubs } from './ConstantCats';
import { FIXED_HEADER_NONE, FIXED_HEADER_FULL, FIXED_HEADER_SUB, FIXED_HEADER_SEARCH } from './ConstantUI';

const DISTANCE_OVER_SUBHEADER = 80;
const EMPTY_FUNCTION = () => {};

function findCatSlugFromUrl(pathname = '') {
  const regex = /^\/(news|columnists|projects|trending|search)?\/?(\w+)?\/?(\w+)?/;
  const matches = pathname.match(regex);
  let target = null;

  if (matches) {
    if (matches[1] && matches[1] === 'columnists') {
      target = 'celebrity_area';
    } else if (matches[2] && matches[2] === 'cat') {
      target = matches[3];
    } else if (matches[1]) {
      target = matches[1];
    }
  }

  return CategoryMappingWithSubs[target] && CategoryMappingWithSubs[target].parentId !== 0
    ? CategoryMappingWithSubs[target].parentSlug
    : target;
}

function renderNavs(channel, navs, newsBaseUrl, request) {
  return navs.map((nav, idx) => {
    const className = nav.title === channel ? classNames(styles.active, 'theme-active') : '';

    /* eslint-disable react/no-array-index-key */
    return <SubMenu key={idx} className={className} {...nav} newsBaseUrl={newsBaseUrl} request={request} />;
    /* eslint-enable react/no-array-index-key */
  });
}

class Header extends PureComponent {
  static propTypes = {
    channel: PropTypes.string.isRequired,
    location: locationShape.isRequired,
    stickySubHeader: PropTypes.bool,
    stickySearchHeader: PropTypes.bool,
    fixedHeaderType: PropTypes.oneOf([FIXED_HEADER_NONE, FIXED_HEADER_FULL, FIXED_HEADER_SUB, FIXED_HEADER_SEARCH])
      .isRequired,
    toggleFixedHeader: PropTypes.func.isRequired,
    newsBaseUrl: PropTypes.string.isRequired,
    request: requestType,
    navs: navsType.isRequired,
    catNavs: catNavsType,
  };

  static defaultProps = {
    stickySubHeader: false,
    stickySearchHeader: false,
    fixedHeaderType: FIXED_HEADER_NONE,
    newsBaseUrl: '',
    request: undefined,
    catNavs: undefined,
    toggleFixedHeader: EMPTY_FUNCTION,
  };

  componentDidMount() {
    this.props.toggleFixedHeader(FIXED_HEADER_FULL);
    if (this.props.stickySubHeader) {
      window.addEventListener('scroll', this.scrollHandler);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scrollHandler);
  }

  scrollHandler = () => {
    const scrollY = 'scrollY' in window ? window.scrollY : document.documentElement.scrollTop;
    const sticky = this.props.stickySubHeader;

    if (sticky) {
      if (scrollY >= DISTANCE_OVER_SUBHEADER && this.props.fixedHeaderType !== FIXED_HEADER_SUB) {
        this._setNextState({ fixedHeaderType: FIXED_HEADER_SUB });
      } else if (scrollY < DISTANCE_OVER_SUBHEADER && this.props.fixedHeaderType === FIXED_HEADER_SUB) {
        this._setNextState({ fixedHeaderType: FIXED_HEADER_FULL });
      }
    } else if (this.props.stickySearchHeader) {
      this._setNextState({ fixedHeaderType: FIXED_HEADER_SEARCH });
    } else if (scrollY >= DISTANCE_OVER_SUBHEADER) {
      this._setNextState({ fixedHeaderType: FIXED_HEADER_NONE });
    } else if (scrollY < DISTANCE_OVER_SUBHEADER) {
      this._setNextState({ fixedHeaderType: FIXED_HEADER_FULL });
    }
  };

  _setNextState = state => {
    if (this._setNextStateAnimationFrameId) {
      raf.cancel(this._setNextStateAnimationFrameId);
    }

    this._setNextStateAnimationFrameId = raf(() => {
      this._setNextStateAnimationFrameId = null;
      this.props.toggleFixedHeader(state.fixedHeaderType);
    });
  };

  renderSubHeader(isFixed = false) {
    const { fixedHeaderType } = this.props;

    // .js-* className is for e2e test
    let subHeaderClass = isFixed
      ? classNames(styles['sub-header'], styles.fixed, 'theme-sub-header', 'theme-fixed')
      : classNames(styles['sub-header'], 'theme-sub-header');

    if (fixedHeaderType === FIXED_HEADER_SEARCH) {
      if (isFixed) {
        subHeaderClass = classNames(subHeaderClass, styles.hide, 'theme-hide', 'js-hide');
      }
    } else {
      subHeaderClass =
        fixedHeaderType === (isFixed ? FIXED_HEADER_SUB : FIXED_HEADER_FULL)
          ? subHeaderClass
          : classNames(subHeaderClass, styles.hide, 'theme-hide', 'js-hide');
    }

    subHeaderClass = classNames(subHeaderClass, 'js-header-sub-header');

    return (
      <div className={subHeaderClass}>
        <nav className={classNames(styles['cat-menu'], 'theme-cat-menu')}>{this.renderCatMenu()}</nav>
      </div>
    );
  }

  renderCatMenu() {
    const { catNavs, location } = this.props;
    const activeCatSlug = findCatSlugFromUrl(location.pathname);

    return catNavs.map((nav, idx) => {
      return (
        <SubNavItem
          key={idx} // eslint-disable-line react/no-array-index-key
          url={nav.url}
          title={nav.title}
          isNew={nav.name === 'search'}
          isActive={nav.name === activeCatSlug}
          subItems={nav.subItems}
          external={nav.external}
        />
      );
    });
  }

  renderSearch() {
    return (
      <form
        acceptCharset="UTF-8"
        action="https://so.cnyes.com/cnyessearch.aspx"
        className={styles['header-search']}
        target="_blank"
      >
        <input type="hidden" name="cx" value="015486011444191663508:8ijuvgfglaq" />
        <input
          type="hidden"
          name="other"
          value=""
          ref={ref => {
            this.otherInput = ref;
          }}
        />
        <input type="hidden" name="ie" value="UTF-8" />
        <input type="hidden" name="ga" value="nav" />
        <input
          name="q"
          placeholder="請輸入關鍵詞"
          onChange={e => {
            this.otherInput.value = e.target.value;
          }}
        />
        <button type="submit" />
      </form>
    );
  }

  render() {
    const { catNavs, channel, navs, newsBaseUrl, request } = this.props;

    return (
      <div id={styles.wrapper} className={classNames('theme-wrapper', 'theme-header')}>
        <header className={classNames(styles['main-header'], 'theme-main-header')}>
          <div className={styles['header-menu']}>
            <span className={styles['logo-wrapper']}>
              <a href="http://www.cnyes.com/" className={styles.logo} />
              <Link to="/" className={styles['channel-label']}>
                {channel}
              </Link>
            </span>
            <span className={styles.actions}>
              <ul className={styles['user-nav']}>
                <li>
                  <a href="https://www.facebook.com/cnYES/" target="_blank" rel="noopener noreferrer">
                    粉絲團
                  </a>
                </li>
              </ul>
              {this.renderSearch()}
            </span>
          </div>
          <nav>{renderNavs(channel, navs, newsBaseUrl, request)}</nav>
        </header>
        {catNavs &&
          catNavs.length && (
            <div
              className={classNames(
                styles['subheader-wrapper'],
                this.props.fixedHeaderType === FIXED_HEADER_SUB && styles.fixed
              )}
            >
              {this.renderSubHeader(true)}
              {this.renderSubHeader(false)}
            </div>
          )}
      </div>
    );
  }
}

export default Header;
