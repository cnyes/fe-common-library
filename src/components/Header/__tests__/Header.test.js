/* eslint-disable import/first,func-names,global-require,prefer-arrow-callback */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Header from '../Header';
import { navs, catNavs } from '../config/fund/FundConfig';

describe('<Header/>', function() {
  let makeSubject;
  let Link;

  beforeEach(() => {
    jest.resetModules();

    const testParams = {
      catNavs,
      channel: '基金',
      location: {
        pathname: '/news/cat/headline',
        search: '',
        state: {},
        action: '',
        key: '',
      },
      navs,
      newsBaseUrl: '',
      toggleFixedHeader: jest.fn(),
    };

    Link = jest.fn(() => <div className="mock-link" />);
    makeSubject = (params = {}) => {
      const props = {
        ...testParams,
        ...params,
      };

      return mount(<Header {...props} />);
    };
  });

  it('should match snapshot', () => {
    const subject = makeSubject();

    expect(subject.find('SubMenu').length).toEqual(navs.length);
    expect(subject.find('SubNavItem').length).toEqual(catNavs.length * 2);

    expect(toJson(subject)).toMatchSnapshot();
  });

  describe('catNavs', () => {
    it('should be nice if it is undefined', () => {
      const subject = makeSubject({
        catNavs: undefined,
      });

      expect(subject.find('SubNavItem').length).toEqual(0);
    });

    it('should be nice if it has no data', () => {
      const subject = makeSubject({
        catNavs: [],
      });

      expect(subject.find('SubNavItem').length).toEqual(0);
    });

    it('should be nice if it has data', () => {
      const subject = makeSubject();

      expect(subject.find('SubNavItem').length).not.toEqual(0);
    });
  });

  describe('displayChannelName', () => {
    it('should be nice if it is true', () => {
      const subject = makeSubject();
      const channelName = subject.props().channel;

      expect(subject.props().displayChannelName).toEqual(true);
      expect(subject.find('.logo-wrapper').html()).toContain(channelName);
    });

    it('should be nice if it is false', () => {
      const subject = makeSubject({
        displayChannelName: false,
      });
      const channelName = subject.props().channelName;

      expect(subject.props().displayChannelName).toEqual(false);
      expect(subject.find('.logo-wrapper').html()).not.toContain(channelName);
    });

    it('should be a Link component', () => {
      const subject = makeSubject({ Link });

      expect(subject.props().Link).toBeDefined();
      expect(subject.find('.logo-wrapper .mock-link').length).toEqual(1);
    });

    it('should be an a tag', () => {
      const subject = makeSubject();

      expect(subject.props().Link).not.toBeDefined();
      expect(subject.find('.logo-wrapper .mock-link').length).toEqual(0);
      expect(subject.find('.logo-wrapper').html()).toContain('channel-label');
    });
  });

  describe('stickySubHeader', () => {
    describe('when true', () => {
      it('should add event listener to window', () => {
        window.addEventListener = jest.fn();
        makeSubject({ stickySubHeader: true });

        expect(window.addEventListener).toHaveBeenCalledTimes(1);
        expect(window.addEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));
      });
    });

    describe('when false', () => {
      it('should add event listener to window', () => {
        window.addEventListener = jest.fn();
        makeSubject({ stickySubHeader: false });

        expect(window.addEventListener).toHaveBeenCalledTimes(0);
      });
    });
  });
});
