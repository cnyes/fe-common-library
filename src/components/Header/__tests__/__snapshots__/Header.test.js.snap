// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header/> should match snapshot 1`] = `"<div id=\\"cnyes-header-wrapper\\" class=\\"theme-wrapper theme-header\\"><header class=\\"main-header theme-main-header\\"><div class=\\"header-menu\\"><span class=\\"logo-wrapper\\"><a href=\\"http://www.cnyes.com/\\" class=\\"logo\\"></a><a href=\\"/\\" class=\\"channel-label\\">基金</a></span><span class=\\"actions\\"><ul class=\\"user-nav\\"><li><a href=\\"https://www.facebook.com/cnYES/\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">粉絲團</a></li></ul><form accept-charset=\\"UTF-8\\" action=\\"https://so.cnyes.com/cnyessearch.aspx\\" class=\\"header-search\\" target=\\"_blank\\"><input type=\\"hidden\\" name=\\"cx\\" value=\\"015486011444191663508:8ijuvgfglaq\\"><input type=\\"hidden\\" name=\\"other\\" value=\\"\\"><input type=\\"hidden\\" name=\\"ie\\" value=\\"UTF-8\\"><input type=\\"hidden\\" name=\\"ga\\" value=\\"nav\\"><input name=\\"q\\" placeholder=\\"請輸入關鍵詞\\"><button type=\\"submit\\"></button></form></span></div><nav><div class=\\"\\"><a href=\\"http://www.cnyes.com/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 21 -->首頁<!-- /react-text --></a></div><div class=\\"\\"><a href=\\"http://news.cnyes.com/\\" class=\\"submenu-title\\"><!-- react-text: 24 -->新聞<!-- /react-text --></a></div><div class=\\"\\"><a href=\\"http://www.cnyes.com/twstock/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 27 -->台股<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://www.cnyes.com/twstock/index.htm\\" title=\\"台指期\\">台指期</a><a href=\\"http://www.cnyes.com/twstock/index.htm\\" title=\\"選擇權\\">選擇權</a><a href=\\"http://www.cnyes.com/warrant/tw/index.htm\\" title=\\"權證\\">權證</a><a href=\\"http://www.cnyes.com/presh/index.htm\\" title=\\"興櫃\\">興櫃</a><a href=\\"http://www.cnyes.com/pre/index.htm\\" title=\\"未上市\\">未上市</a></nav></div></div><div class=\\"\\"><a href=\\"https://stock.cnyes.com/\\" class=\\"submenu-title\\"><!-- react-text: 38 -->股市Talk<!-- /react-text --></a></div><div class=\\"\\"><a href=\\"http://www.cnyes.com/forex/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 41 -->外匯<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"/news/cat/forex\\" title=\\"新聞\\">新聞</a><a href=\\"http://traderoom.cnyes.com/global/forex.aspx\\" title=\\"即時報價\\">即時報價</a><a href=\\"http://www.cnyes.com/forex/forex_index.aspx\\" title=\\"看盤室\\">看盤室</a><a href=\\"http://www.cnyes.com/forex/major.aspx\\" title=\\"主要匯價\\">主要匯價</a><a href=\\"http://www.cnyes.com/forex/forex_list.aspx\\" title=\\"交叉匯率\\">交叉匯率</a><a href=\\"http://www.cnyes.com/forex/twd.aspx\\" title=\\"新台幣\\">新台幣</a><a href=\\"http://www.cnyes.com/forex/rmb.htm\\" title=\\"人民幣\\">人民幣</a><a href=\\"http://www.cnyes.com/forex/jpy.htm\\" title=\\"日元\\">日元</a><a href=\\"http://www.cnyes.com/forex/eur.htm\\" title=\\"歐元\\">歐元</a></nav></div></div><div class=\\"\\"><a href=\\"http://www.cnyes.com/global/IndexImmediateQuotedPrice/\\" class=\\"submenu-title\\"><!-- react-text: 56 -->全球市場<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://www.cnyes.com/global/IndexImmediateQuotedPrice/\\" title=\\"國際股\\">國際股</a><a href=\\"http://www.cnyes.com/USASTOCK/index.htm\\" title=\\"美股\\">美股</a><a href=\\"http://www.cnyes.com/hkstock/index.htm\\" title=\\"港股\\">港股</a><a href=\\"http://www.cnyes.com/shstock/index.htm\\" title=\\"滬深股\\">滬深股</a><a href=\\"http://www.cnyes.com/JP/index.htm\\" title=\\"日股\\">日股</a><a href=\\"http://www.cnyes.com/futures/index.htm\\" title=\\"國際期貨\\">國際期貨</a><a href=\\"http://www.cnyes.com/bond/index.htm\\" title=\\"債券\\">債券</a><a href=\\"http://www.cnyes.com/gold/index.htm\\" title=\\"黃金\\">黃金</a><a href=\\"http://www.cnyes.com/CentralBank/index.htm\\" title=\\"全球央行\\">全球央行</a><a href=\\"http://www.cnyes.com/economy/indicator/Page/schedule.aspx\\" title=\\"經濟指標\\">經濟指標</a><a href=\\"http://money.cnyes.com/StockQ.aspx\\" title=\\"StockQ\\">StockQ</a></nav></div></div><div class=\\"active theme-active\\"><a href=\\"http://fund.cnyes.com/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 73 -->基金<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://fund.cnyes.com/MyFunds.aspx\\" title=\\"我的基金\\">我的基金</a><a href=\\"http://fund.cnyes.com/Search.aspx\\" title=\\"基金搜尋\\">基金搜尋</a><a href=\\"http://fund.cnyes.com/ranking/index.htm\\" title=\\"基金排行\\">基金排行</a><a href=\\"http://fund.cnyes.com/Fixedincome/\\" title=\\"債券專區\\">債券專區</a><a href=\\"http://fund.cnyes.com/dividend/index.htm\\" title=\\"配息專區\\">配息專區</a><a href=\\"http://fund.cnyes.com/report/report/index.htm\\" title=\\"研究報告\\">研究報告</a><a href=\\"https://www.fundsyes.com/?utm_source=cnyes&amp;utm_medium=index_menu\\" title=\\"基金交易平台\\">基金交易平台</a></nav></div></div><div class=\\"\\"><span class=\\"submenu-title\\"><!-- react-text: 86 -->交易<!-- /react-text --><span class=\\"with-arrow\\"></span></span><div class=\\"popup theme-popup\\"><nav class=\\"link-wrapper theme-link-wrapper\\"><a href=\\"https://www.fundsyes.com/Index.aspx?utm_source=cnyes&amp;utm_medium=channel_mainpage\\" title=\\"基金交易\\">基金交易</a></nav><nav class=\\"news-list theme-news-list\\"><h5>鉅亨基金交易平台</h5><a href=\\"https://www.fundsyes.com/Login.aspx?utm_source=cnyes&amp;utm_medium=channel_login\\" title=\\"交易登入\\">交易登入</a><a href=\\"https://www.fundsyes.com/launch/fund_index.htm?utm_source=cnyes&amp;utm_medium=channel_newuser\\" title=\\"免費開戶\\">免費開戶</a><a href=\\"https://www.fundsyes.com/Guide/Expenses.aspx?utm_source=cnyes&amp;utm_medium=channel_preferential\\" title=\\"最新優惠\\">最新優惠</a><a href=\\"https://www.fundsyes.com/Tool/SuggestionList.aspx?utm_source=cnyes&amp;utm_medium=channel_radar\\" title=\\"投資雷達\\">投資雷達</a></nav></div></div><div class=\\"\\"><a href=\\"http://www.cnyes.com/money/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 99 -->理財<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://www.cnyes.com/money/BankService.aspx\\" title=\\"銀行服務\\">銀行服務</a><a href=\\"http://cnyes.com/mortgage\\" title=\\"房貸\\">房貸</a><a href=\\"http://www.cnyes.com/money/fundptcalcu.aspx\\" title=\\"投資分析\\">投資分析</a><a href=\\"http://www.cnyes.com/money/livelihoodPrice.htm\\" title=\\"民生物價\\">民生物價</a><a href=\\"http://www.cnyes.com/money/BankCalculation.aspx\\" title=\\"試算工具\\">試算工具</a><a href=\\"/news/cat/tw_money\\" title=\\"理財新聞\\">理財新聞</a><a href=\\"http://mag.cnyes.com/Column/financial/index.shtml\\" title=\\"理財雜誌\\">理財雜誌</a><a href=\\"http://www.cnyes.com/fixedincome/index.htm\\" title=\\"固定收益\\">固定收益</a><a href=\\"http://www.cnyes.com/insurance/index.htm\\" title=\\"保險\\">保險</a><a href=\\"http://www.cnyes.com/leases/index.htm\\" title=\\"租賃\\">租賃</a><a href=\\"http://www.cnyes.com/globalinvest/GlobalStock.htm\\" title=\\"股市地圖\\">股市地圖</a></nav></div></div><div class=\\"\\"><a href=\\"http://house.cnyes.com/global/\\" class=\\"submenu-title\\"><!-- react-text: 116 -->海外房產<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://house.cnyes.com/global/malaysia/default.aspx\\" title=\\"馬來西亞\\">馬來西亞</a><a href=\\"http://house.cnyes.com/global/japan/\\" title=\\"日本\\">日本</a><a href=\\"http://house.cnyes.com/global/thailand/\\" title=\\"泰國\\">泰國</a><a href=\\"http://house.cnyes.com/global/Cambodia/\\" title=\\"柬埔寨\\">柬埔寨</a><a href=\\"http://house.cnyes.com/global/Australia/\\" title=\\"澳洲\\">澳洲</a><a href=\\"http://house.cnyes.com/global/American/\\" title=\\"美國\\">美國</a><a href=\\"http://house.cnyes.com/\\" title=\\"台灣\\">台灣</a><a href=\\"http://house.cnyes.com/News/\\" title=\\"房地產新聞\\">房地產新聞</a></nav></div></div><div class=\\"\\"><a href=\\"http://blog.cnyes.com/\\" class=\\"submenu-title\\"><!-- react-text: 130 -->部落格<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\" style=\\"width: 120px; height: auto; overflow: visible;\\"><nav class=\\"link-wrapper theme-link-wrapper\\" style=\\"height: auto;\\"><a href=\\"http://blog.cnyes.com/Excerpt.aspx\\" title=\\"名家專區\\">名家專區</a><a href=\\"http://blog.cnyes.com/SearchBlog.aspx?Purpose=Hit&amp;ga=nav\\" title=\\"人氣排行榜\\">人氣排行榜</a><a href=\\"http://blog.cnyes.com/WorldFresh.aspx\\" title=\\"最新文章\\">最新文章</a></nav></div></div><div class=\\"\\"><a href=\\"http://theme.cnyes.com/fund/index.html\\" class=\\"submenu-title\\"><!-- react-text: 139 -->主題投資<!-- /react-text --></a></div></nav></header><div class=\\"subheader-wrapper\\"><div class=\\"sub-header fixed theme-sub-header theme-fixed hide theme-hide js-hide js-header-sub-header\\"><nav class=\\"cat-menu theme-cat-menu\\"><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/MyFunds.aspx\\" title=\\"我的基金\\">我的基金</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/myInvestment/myInvestment.aspx?ga=nav\\" title=\\"投資組合分析\\">投資組合分析</a><a href=\\"/calcmyfundHTML5.aspx?ga=nav\\" title=\\"基金透視鏡 \\">基金透視鏡 </a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"is-new\\" href=\\"/search/\\" title=\\"基金搜尋\\">基金搜尋</a></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/ranking/index.htm?ga=nav\\" title=\\"基金排行\\">基金排行</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/sector-ranking/index.htm?ga=nav\\" title=\\"組別分類排行\\">組別分類排行</a><a href=\\"/sector-ranking-yearly/index.htm?ga=nav\\" title=\\"組別年度績效排行\\">組別年度績效排行</a><a href=\\"/popular/index.htm?ga=nav\\" title=\\"熱門基金排行\\">熱門基金排行</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/Fixedincome/\\" title=\\"債券專區\\">債券專區</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/Fixedincome/index.aspx\\" title=\\"債券基金排行\\">債券基金排行</a><a href=\\"/Fixedincome/search.aspx\\" title=\\"債券基金搜尋器\\">債券基金搜尋器</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"\\" href=\\"/dividend/index.htm\\" title=\\"配息專區\\">配息專區</a></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/report/report/index.htm\\" title=\\"研究報告\\">研究報告</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/report/manager/index.htm?ga=nav\\" title=\\"經理人訪談\\">經理人訪談</a><a href=\\"http://campaign.cnyes.com/FidelityAsiaBond/?ga=nav\\" title=\\"亞債新視野\\">亞債新視野</a><a href=\\"/report/MorningStar/index.htm?ga=nav\\" title=\\"晨星專欄\\">晨星專欄</a><a href=\\"/report/cnyes/index.htm?ga=nav\\" title=\\"鉅亨投資雷達\\">鉅亨投資雷達</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"\\" href=\\"https://www.fundsyes.com/?utm_source=cnyes&amp;utm_medium=fundchannel_menu2\\" title=\\"基金交易平台\\">基金交易平台</a></span></nav></div><div class=\\"sub-header theme-sub-header hide theme-hide js-hide js-header-sub-header\\"><nav class=\\"cat-menu theme-cat-menu\\"><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/MyFunds.aspx\\" title=\\"我的基金\\">我的基金</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/myInvestment/myInvestment.aspx?ga=nav\\" title=\\"投資組合分析\\">投資組合分析</a><a href=\\"/calcmyfundHTML5.aspx?ga=nav\\" title=\\"基金透視鏡 \\">基金透視鏡 </a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"is-new\\" href=\\"/search/\\" title=\\"基金搜尋\\">基金搜尋</a></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/ranking/index.htm?ga=nav\\" title=\\"基金排行\\">基金排行</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/sector-ranking/index.htm?ga=nav\\" title=\\"組別分類排行\\">組別分類排行</a><a href=\\"/sector-ranking-yearly/index.htm?ga=nav\\" title=\\"組別年度績效排行\\">組別年度績效排行</a><a href=\\"/popular/index.htm?ga=nav\\" title=\\"熱門基金排行\\">熱門基金排行</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/Fixedincome/\\" title=\\"債券專區\\">債券專區</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/Fixedincome/index.aspx\\" title=\\"債券基金排行\\">債券基金排行</a><a href=\\"/Fixedincome/search.aspx\\" title=\\"債券基金搜尋器\\">債券基金搜尋器</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"\\" href=\\"/dividend/index.htm\\" title=\\"配息專區\\">配息專區</a></span><span class=\\"cat-nav-item\\"><a class=\\"with-arrow\\" href=\\"/report/report/index.htm\\" title=\\"研究報告\\">研究報告</a><nav class=\\"cat-nav-sub-item\\"><a href=\\"/report/manager/index.htm?ga=nav\\" title=\\"經理人訪談\\">經理人訪談</a><a href=\\"http://campaign.cnyes.com/FidelityAsiaBond/?ga=nav\\" title=\\"亞債新視野\\">亞債新視野</a><a href=\\"/report/MorningStar/index.htm?ga=nav\\" title=\\"晨星專欄\\">晨星專欄</a><a href=\\"/report/cnyes/index.htm?ga=nav\\" title=\\"鉅亨投資雷達\\">鉅亨投資雷達</a></nav></span><span class=\\"cat-nav-item\\"><a class=\\"\\" href=\\"https://www.fundsyes.com/?utm_source=cnyes&amp;utm_medium=fundchannel_menu2\\" title=\\"基金交易平台\\">基金交易平台</a></span></nav></div></div></div>"`;
